tests/phpunit/tests/fonts/font-library/wpFontsDir.php

```diff
@@ -69,4 +76,113 @@ function set_new_values( $defaults ) {
+	}
+
+	public function test_should_create_fonts_dir_in_uploads_when_fails_in_wp_content() {
+		// Set the expected results.
+		$upload_dir = wp_upload_dir();
+		$expected   = array(
+			'path'    => path_join( $upload_dir['basedir'], 'fonts' ),
+			'url'     => $upload_dir['baseurl'] . '/fonts',
+			'subdir'  => '',
+			'basedir' => path_join( $upload_dir['basedir'], 'fonts' ),
+			'baseurl' => $upload_dir['baseurl'] . '/fonts',
+			'error'   => false,
+		);
+
+		$this->create_fake_file_to_avoid_dir_creation( static::$dir_defaults['path'] );
+		$this->assertFileExists( static::$dir_defaults['path'] );
+
+		$font_dir = wp_get_font_dir();
+
+		$this->assertDirectoryDoesNotExist( static::$dir_defaults['path'], 'The `wp-content/fonts` directory should not exist.' );
+		$this->assertDirectoryExists( $font_dir['path'], 'The `uploads/fonts` directory should exist.' );
+		$this->assertSame( $expected, $font_dir, 'The `fonts` directory should be a subdir in the `uploads` directory.' );
+	}
+
+	public function test_should_return_error_if_unable_to_create_fonts_dir_in_uploads() {
+		// Disallow the creation of the `wp-content/fonts` directory.
+		$this->create_fake_file_to_avoid_dir_creation( static::$dir_defaults['path'] );
+		$this->assertFileExists( static::$dir_defaults['path'] );
+
+		// Disallow the creation of the `uploads/fonts` directory.
+		$upload_dir       = wp_upload_dir();
+		$font_upload_path = path_join( $upload_dir['basedir'], 'fonts' );
+		$this->create_fake_file_to_avoid_dir_creation( $font_upload_path );
+		$this->assertFileExists( $font_upload_path );
+
+		$expected = array(
+			'path'    => $font_upload_path,
+			'url'     => $upload_dir['baseurl'] . '/fonts',
+			'subdir'  => '',
+			'basedir' => $font_upload_path,
+			'baseurl' => $upload_dir['baseurl'] . '/fonts',
+			'error'   => 'Unable to create directory wp-content/uploads/fonts. Is its parent directory writable by the server?',
+		);
+
+		$font_dir = wp_get_font_dir();
+
+		$this->assertDirectoryDoesNotExist( $font_upload_path, 'The `uploads/fonts` directory should not exist.' );
+		$this->assertSame( $expected, $font_dir, 'As /wp-content/uplods/fonts is not writable the error key should be populated with an error message.' );
+	}
+
+	private function remove_font_paths() {
+		$paths = array(
+			path_join( WP_CONTENT_DIR, 'fonts' ),
+			path_join( WP_CONTENT_DIR, 'custom_dir' ),
+			path_join( WP_CONTENT_DIR, 'uploads/fonts' ),
+		);
+
+		foreach ( $paths as $path ) {
+			if ( ! is_dir( $path ) ) {
+				if ( file_exists( $path ) ) {
+					unlink( $path );
+				}
+			} else {
+				$this->rmdir( $path );
+				rmdir( $path );
+			}
+		}
+	}
+
+	/**
+	 * A placeholder "fake" file at $path triggers `wp_mkdir_p()` to fail into the `file_exists()` bail out, causing `is_dir()` to  return `false`.
+	 * This effectively makes $path unwritable.
+	 */
+	private function create_fake_file_to_avoid_dir_creation( $path ) {
+		file_put_contents(
+			$path,
+			'fake file'
+		);
+	}
```

Comment:
These two methods will cause problems for most local development environments, including the docker container provided in this repo.

Typically the test suite and the development environment share content directories so developers running these environments may have existing font folders as a result of their manual testing.

During my tests the fonts from manual testing were deleted during these test runs. 

To reproduce using the repos docker environment:

1. Run commands to bootstrap the environment
   ```
   npm install
   npm run build:dev
   npm run env:start
   npm run env:install
   ```
2. Visit http://localhost:8889/wp-admin/site-editor.php?canvas=edit
3. Opens Styles > Typography > Manager > Install Fonts & Install some fonts via google
4. Apply the font to the main header in 2024
5. Save and exit the site editor
6. Observe the folder `src/wp-content/fonts` has been created
7. Run this test suite `npm run test:php -- --filter Tests_Fonts_WpFontDir`
8. Observe the folder `src/wp-content/fonts` has been created
9. Visit the front end of the site 
10. Observe the font download fails with a 404 error

Reply:
> Typically the test suite and the development environment share content directories so developers running these environments may have existing font folders as a result of their manual testing.

I think this is unfortunate and unexpected, right? The unit tests environment should run in a reproducible environment and not depend on or produce 'side effects.' It seems like if other parts of the workflow unrelated to unit tests are affected by the changes made by running the unit tests, the workflow is deficient, and it should be improved instead of degrading the unit tests.

Reply:
> Check if the fonts folder already exists

+1. If the /fonts dir already exists (at any tested location) it can be used for testing but shouldn't be deleted.

Reply:
The approach seems sound, though I'd do the same for `wp-content/uploads/fonts`. This may also already exist for manual testing purposes, and the test suite shouldn't delete it.

Ideally, we would be using a tests-only fonts directory to avoid affecting the actual site's filesystem in the tests and also preventing possible filename clashes, etc. Alternatively, if we continue to use the actual site's filesystem, maybe store the existence _and_ pre-test contents of the directories. This means we could accurately compare the contents before/after the function under test has run, and therefore detect unexpected creation/deletion of files in those directories, plus delete only the files created by the tests.